deeppavlov/dataset_readers/md_yaml_dialogs_reader.py

```diff
@@ -0,0 +1,388 @@
+                  f"domain_knowledge={domain_knowledge}, "
+                  f"intent2slots2text={intent2slots2text}, "
+                  f"slot_name2text2value={slot_name2text2value}")
+
+        return gobot_formatted_stories
+
+    @classmethod
+    def _choose_slots_for_whom_exists_text(cls, intent2slots2text, slots_actual_values, user_action):
+        possible_keys = [k for k in intent2slots2text.keys() if user_action in k]
+        possible_keys = possible_keys + [user_action]
+        possible_keys = sorted(possible_keys, key=lambda action_s: action_s.count('+'))
+        for possible_action_key in possible_keys:
+            if intent2slots2text[possible_action_key].get(slots_actual_values):
+                slots_used_values = slots_actual_values
+                slots_to_exclude = []
+                return slots_to_exclude, slots_used_values, possible_action_key
+            else:
+                slots_lazy_key = set(e[0] for e in slots_actual_values)
+                slots_lazy_key -= {"intent"}
+                fake_keys = []
+                for known_key in intent2slots2text[possible_action_key].keys():
+                    if slots_lazy_key.issubset(set(e[0] for e in known_key)):
+                        fake_keys.append(known_key)
+                        break
+                fake_key = None
+                if fake_keys:
+                    fake_key = sorted(fake_keys,
+                                      key=lambda elem: (len(set(slots_actual_values) ^ set(elem)),
+                                                        len([e for e in elem if e[0] not in slots_lazy_key])))[0]
+
+                    slots_to_exclude = [e[0] for e in fake_key if e[0] not in slots_lazy_key]
+                    slots_used_values = fake_key
+                    return slots_to_exclude, slots_used_values, possible_action_key
+
+        raise KeyError("no possible NLU candidates found")
+
+    @classmethod
+    def _clarify_slots_values(cls, slot_name2text2value, slots_dstc2formatted):
+        slots_key = []
+        for slot_name, slot_value in slots_dstc2formatted:
+            if slot_value in slot_name2text2value.get(slot_name, {}).keys():
+                slot_actual_value = slot_name2text2value[slot_name][slot_value]
+            else:
+                slot_actual_value = slot_value
+            slots_key.append((slot_name, slot_actual_value))
+        slots_key = tuple(sorted(slots_key))
+        return slots_key
+
+    @classmethod
+    def _parse_user_intent(cls, line):
+        intent = line.strip('*').strip()
+        if '{' not in intent:
+            intent = intent + "{}"  # the prototypical intent is "intent_name{slot1: value1, slotN: valueN}"
+        user_action, slots_info = intent.split('{', 1)
+        slots_info = json.loads('{' + slots_info)
+        slots_dstc2formatted = [[slot_name, slot_value] for slot_name, slot_value in slots_info.items()]
+        return user_action, slots_dstc2formatted
+
+    @classmethod
```

Comment:
staticmethod, remove cls, add `slots_li` typing

Reply:
done

Reply:
I don't think so. You made `slots_li=SLOT2VALUE_PAIRS_TUPLE` instead of `slits_li: Optional[SLOT2VALUE_PAIRS_TUPLE] = None`

Reply:
No, Optional is necessary https://docs.python.org/3/library/typing.html#typing.Optional, despite it looks in this case quite ugly

Reply:
done